shared/ssz/encode.go

```diff
@@ -0,0 +1,132 @@
+package ssz
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"reflect"
+)
+
+// TODOs for this PR:
+// - Review all error handling
+// - Use customized error types to avoid error string everywhere!
+// - Add check for input sizes! (2 << 32)
+// - Use constant to replace hard coding of number 4
+
+// TODOs for later PR:
+// - Add support for more types
+
+// Encode TODO add comments
+func Encode(w io.Writer, val interface{}) error {
+	eb := &encbuf{}
+	if err := eb.encode(val); err != nil {
+		return err
+	}
+	return eb.toWriter(w)
+}
+
+type encbuf struct {
+	str []byte
+}
+
+func (w *encbuf) encode(val interface{}) error {
+	rval := reflect.ValueOf(val)
+	encDec, err := getEncoderDecoderForType(rval.Type())
+	if err != nil {
+		return err
+	}
+	return encDec.encoder(rval, w)
+}
+
+func (w *encbuf) toWriter(out io.Writer) error {
+	if _, err := out.Write(w.str); err != nil {
+		return err
+	}
+	return nil
+}
+
+func makeEncoder(typ reflect.Type) (encoder, error) {
+	kind := typ.Kind()
+	switch {
+	case kind == reflect.Uint8:
+		return encodeUint8, nil
+	case kind == reflect.Uint16:
+		return encodeUint16, nil
+	case kind == reflect.Slice && typ.Elem().Kind() == reflect.Uint8:
+		return encodeBytes, nil
+	case kind == reflect.Slice:
+		return makeSliceEncoder(typ)
+	case kind == reflect.Struct:
+		return makeStructEncoder(typ)
+	default:
```

Comment:
Correct me if I'm wrong. Do you plan on supporting a custom encoder interface (similar to this: https://github.com/ethereum/go-ethereum/blob/master/rlp/encode.go#L350 and for this interface to provide  `Encode` and `EncodingSize` methods?

Reply:
Yes exactly.